tensorboard/uploader/uploader.py

```diff
@@ -262,41 +257,50 @@ def build_requests(self, run_to_events):
             metadata = self._tag_metadata.get(time_series_key)
             if metadata is None:
-                plugin_name = value.metadata.plugin_data.plugin_name
-                if plugin_name == scalar_metadata.PLUGIN_NAME:
-                    metadata = value.metadata
-                else:
-                    metadata = _RequestBuilder._NON_SCALAR_TIME_SERIES
+                metadata = value.metadata
                 self._tag_metadata[time_series_key] = metadata
-            if metadata is _RequestBuilder._NON_SCALAR_TIME_SERIES:
-                next(work_items)
-                continue
-            try:
-                run_proto = runs.get(run_name)
-                if run_proto is None:
-                    run_proto = self._create_run(run_name)
-                    runs[run_name] = run_proto
-                tag_proto = tags.get((run_name, value.tag))
-                if tag_proto is None:
-                    tag_proto = self._create_tag(run_proto, value.tag, metadata)
-                    tags[(run_name, value.tag)] = tag_proto
-                self._create_point(tag_proto, event, value)
+
+            requests = []  # we may assign a generator
+            if value.HasField("metadata") and (
+                value.metadata.plugin_data.plugin_name
+                != metadata.plugin_data.plugin_name
+            ):
+                logger.warning(
+                    "Mismatching plugin names for %s.  Expected %s, found %s.",
+                    time_series_key,
+                    metadata.plugin_data.plugin_name,
+                    value.metadata.plugin_data.plugin_name,
+                )
+            elif (
+                metadata.plugin_data.plugin_name == scalar_metadata.PLUGIN_NAME
+            ):
+                (
+                    requests,
+                    event_was_consumed,
+                ) = self._scalar_request_builder.add_event(
+                    run_name, event, value, metadata
+                )
+            # TODO(nielsene): add Tensor plugin cases here
+            # TODO(soergel): add Graphs blob case here
+
+            for request in requests:
+                yield request
+
+            # If the current event was not consumed, don't advance work_items,
+            # so the event will be retried.
+            if event_was_consumed:
                 next(work_items)
-            except _OutOfSpaceError:
-                # Flush request and start a new one.
-                request_to_emit = self._prune_request()
-                if request_to_emit is None:
-                    raise RuntimeError("Could not make progress uploading data")
-                self._new_request()
-                runs.clear()
-                tags.clear()
-                yield request_to_emit
-
-        final_request = self._prune_request()
-        if final_request is not None:
-            yield final_request
+
+        requests = self._scalar_request_builder.emit_requests()
```

Comment:
I don't think so, because `flush` often (usually?) means putting buffered data somewhere (e.g., disk) _as a side effect_-- not returning the data, as we do here.